src/clippoly/include/clippoly.h

```diff
@@ -0,0 +1,87 @@
+// Copyright 2021 The Manifold Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <clipper2/clipper.h>
+
+#include "clipper2/clipper.offset.h"
+#include "glm/ext/vector_float2.hpp"
+#include "public.h"
+
+namespace C2 = Clipper2Lib;
+
+namespace manifold {
+class Clippoly {
+ public:
+  Clippoly();
+  ~Clippoly();
+  Clippoly(const Clippoly& other);
+  Clippoly& operator=(const Clippoly& other);
+  Clippoly(Clippoly&&) noexcept;
+  Clippoly& operator=(Clippoly&&) noexcept;
+  Clippoly(std::vector<glm::vec2> contour);
+  Clippoly(std::vector<std::vector<glm::vec2>> contours);
+
+  // Shapes
+  static Clippoly Square(glm::vec2 dims, bool center = false);
+  static Clippoly Circle(float radius, int circularSegments);
+
+  // Booleans
+  enum class OpType { Add, Subtract, Intersect, Xor };
+  Clippoly Boolean(const Clippoly& second, OpType op) const;
+  static Clippoly BatchBoolean(const std::vector<Clippoly>& clippolys,
+                               OpType op);
+  Clippoly operator+(const Clippoly&) const;  // Add (Union)
+  Clippoly& operator+=(const Clippoly&);
+  Clippoly operator-(const Clippoly&) const;  // Subtract (Difference)
+  Clippoly& operator-=(const Clippoly&);
+  Clippoly operator^(const Clippoly&) const;  // Intersect
+  Clippoly& operator^=(const Clippoly&);
+
+  // Transformations
+  Clippoly Translate(glm::vec2 v);
+  Clippoly Scale(glm::vec2 s);
+
+  // Path Simplification
+  Clippoly TrimCollinear();
+  Clippoly Simplify(double epsilon = 1e-6);
+  Clippoly RamerDouglasPeucker(double epsilon = 1e-6);
+  Clippoly StripNearEqual(double epsilon = 1e-6);
+  Clippoly StripDuplicates();
+
+  enum class JoinType { Square, Round, Miter };
+  Clippoly Offset(double delta, JoinType jt, double miter_limit = 2.0,
+                  double arc_tolerance = 0.0);
+
+  // Minkowski
+  // NOTE: OpenSCAD does not use Minkowski as is (from Clipper1). Also, they
+  // allow a list patterns (see applyMinkowski). Should read through that and
+  // write these methods such that they can take Clippoly rather than a single
+  // path pattern.
```

Comment:
You can Minkowski a polygon with holes onto a polygon with holes? I can't even imagine that. I honestly never found much use for Minkowski anyway, especially if we already have offset. What's the practical application here?

Reply:
I've never really used minkowski either to be honest, but yea in OpenSCAD minkowski takes a list of polygons (which can have holes) rather than just a pattern like clipper2 does. 

The sticky part API wise is that it a pattern must be given as a singular path rather than a CrossSection, we'd have to have shape constructors outside of CrossSection as well for convenience.

Reply:
There could be a `Contour` class that wraps Clipper2 `PathD` in a similar way to `CrossSection`, but it's of course more duplication (just like there is in Clipper2 between `PathD` and `PathsD`).

I've cooked up a scheme that I think works pretty nicely in my OCaml Clipper2 bindings, where the single and plural paths are wrapped in an [abstract type](https://github.com/geoffder/ocaml-clipper2/blob/6d62f2d385cbd124c672560bdbb6672876c21326/lib/clipper.ml#L20) that allows the same functions to operate on both. In OOP land though, the user experience would be fairly seemless with similar methods on Contour vs CrossSection.

Reply:
OpenSCAD is great, but I don't feel the need to follow their every move. I don't see much value in Minkowski and I don't like the idea of complicating everything to support it (since this seems to be the only place a simple polygon is required). Let's leave it out for now and see if someone comes to us with a powerful use case.